### PR TITLE
Maintenance Information Improvements

### LINKED ISSUE #1288: Maintenance Improvements
**What would you like to be added** / **Why is this needed**:
Screenshot 2022-08-31 at 18 26 49
Make it clearer when the maintenance starts: e.g. in 5 hours / cluster is currently within the maintenance time window and show the actual time / time window in the tooltip

Extended Scope: To highlight it even more (in addition) we could show a hint (e.g. an alert banner at the top) during the maintenance time window

### LINKED ISSUE #1394: Show maintenance status
**What would you like to be added**:
Show last maintenance operation from shoot status
https://github.com/gardener/gardener/pull/6725

**Why is this needed**:

### PR DESCRIPTION
**Pull request by grolu**
**What this PR does / why we need it**:

***Maintenance Start time improvements*** 
Screenshot 2023-01-11 at 11 34 15
Screenshot 2023-01-11 at 07 24 23

***Show Last Maintenance Status***
Screenshot 2023-01-19 at 12 06 49
Screenshot 2023-01-19 at 12 07 26

**Which issue(s) this PR fixes**:
Fixes #1288 Fixes #1394

**Special notes for your reviewer**:

**Release note**:

```other user
The maintenance start time on the shoot detail page is now shown relative to the current time
```
```other user
The last maintenance status is now shown with an icon next to `Maintenance` on the cluster details page. If the last maintenance status is of state `Failed`, this is shown as a warning on the cluster list
```

### CODE REVIEW

**Comment by grolu** [comment]
Please have a look at the texts. Alternatives could be `Maintenance window opens in...` or `Maintenance window begins in...`
Also I'm not sure about the end. Processes may run longer, right? We could also write `Maintenance window closes in...`
WDYT?

File: frontend/src/components/ShootDetails/ShootLifecycleCard.vue [hunk 1]
```diff
@@ -34,10 +34,33 @@ SPDX-License-Identifier: Apache-2.0
...
+            small
+            class="ml-1" />
+          </v-list-item-title>
+          <v-list-item-subtitle>
+            <v-tooltip top>
+              <template v-slot:activator="{ on }">
+                <div class="d-flex align-center pt-1" v-on="on">
+                  <shoot-messages :shoot-item="shootItem" filter="maintenance-constraint" small class="mr-1" />
+                  <span v-if="isInMaintenanceWindow">
+                    Cluster is currently within the maintenance time window<span v-if="nextMaintenanceEndTimestamp">. Maintenance ends <time-string :date-time="nextMaintenanceEndTimestamp" no-tooltip></time-string></span>
```
Comment (petersutter) [inline frontend/src/components/ShootDetails/ShootLifecycleCard.vue:53]:
```suggestion
                    Cluster is currently within the maintenance time window<span v-if="nextMaintenanceEndTimestamp">. The maintenance time window ends <time-string :date-time="nextMaintenanceEndTimestamp" no-tooltip></time-string></span>
```

File: frontend/src/utils/TimeWithOffset.js [hunk 2]
```diff
@@ -4,6 +4,8 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 
+import moment from '@/utils/moment'
+
```
Comment (holgerkoser) [inline frontend/src/utils/TimeWithOffset.js:8]:
```suggestion
```

File: frontend/src/utils/TimeWithOffset.js [hunk 3]
```diff
@@ -34,6 +36,18 @@ class TimeWithOffset {
...
+    let nextTimestamp = moment(`${this.hours}${this.minutes}${this.offsetSign}${this.offsetHours}${this.offsetMinutes}`, 'HHmmZ')
+    if (!nextTimestamp.isValid()) {
+      return
+    }
+    if (nextTimestamp.isBefore(moment())) {
+      nextTimestamp = nextTimestamp.add(1, 'd')
+    }
+
+    return nextTimestamp.toISOString()
+  }
```
Comment (holgerkoser) [inline frontend/src/utils/TimeWithOffset.js:50]:
```suggestion

  nextDate () {
    let date = new Date()
    const now = date.getTime()
    date = new Date(date.toISOString().substring(0, 11) + this.getTimeString() + this.getTimezoneString())
    while (date.getTime() <= now) {
      date.setUTCDate(date.getUTCDate() + 1)
    }
    return date
  }
```

File: frontend/src/components/ShootDetails/ShootLifecycleCard.vue [hunk 4]
```diff
@@ -139,6 +165,36 @@ export default {
...
+      return `End time: ${maintenanceStartTime.toString()}`
+    },
+    nextMaintenanceBeginTimestamp () {
+      const maintenanceStart = get(this.shootMaintenance, 'timeWindow.begin')
+      const maintenanceStartTime = new TimeWithOffset(maintenanceStart)
+      if (!maintenanceStartTime.isValid()) {
+        return
+      }
+
+      return maintenanceStartTime.getNextTimestampISOString()
```
Comment (holgerkoser) [inline frontend/src/components/ShootDetails/ShootLifecycleCard.vue:184]:
```suggestion
      return maintenanceStartTime.nextDate().toISOString()
```

File: frontend/src/components/ShootDetails/ShootLifecycleCard.vue [hunk 5]
```diff
@@ -139,6 +165,36 @@ export default {
...
+    },
+    nextMaintenanceEndTimestamp () {
+      const maintenanceEnd = get(this.shootMaintenance, 'timeWindow.end')
+      const maintenanceEndTime = new TimeWithOffset(maintenanceEnd)
+      if (!maintenanceEndTime.isValid()) {
+        return
+      }
+
+      return maintenanceEndTime.getNextTimestampISOString()
```
Comment (holgerkoser) [inline frontend/src/components/ShootDetails/ShootLifecycleCard.vue:193]:
```suggestion
      return maintenanceEndTime.nextDate().toISOString()
```

File: frontend/tests/unit/utils.TimeWithOffset.spec.js [hunk 6]
```diff
@@ -35,5 +35,11 @@ describe('utils', () => {
       expect(timeWithOffset.isValid()).toBe(true)
       expect(timeWithOffset.getTimezoneString({ colon: false })).toBe('+0102')
     })
+
+    it('should return next timestamp iso string', () => {
+      const timeWithOffset = new TimeWithOffset('01:02:03+04:05')
+      expect(timeWithOffset.isValid()).toBe(true)
+      expect(timeWithOffset.getNextTimestampISOString().length).toBe(24)
+    })
```
Comment (holgerkoser) [inline frontend/tests/unit/utils.TimeWithOffset.spec.js:43]:
```suggestion

    describe('#nextDate', () => {
      beforeEach(() => {
        jest.useFakeTimers()
      })

      afterEach(() => {
        jest.useRealTimers()
      })

      it('should return next timestamp iso string', () => {
        const timeWithOffset = new TimeWithOffset('01:02+04:05')
        expect(timeWithOffset.isValid()).toBe(true)
        expect(timeWithOffset.nextDate().toISOString()).toMatch(/T20:57:00.000Z$/)
        jest.setSystemTime(new Date('2023-01-01T20:56:59Z'))
        expect(timeWithOffset.nextDate().getUTCDate()).toBe(1)
        jest.setSystemTime(new Date('2023-01-01T20:57:00Z'))
        expect(timeWithOffset.nextDate().getUTCDate()).toBe(2)
        jest.setSystemTime(new Date('2023-01-01T20:57:01Z'))
        expect(timeWithOffset.nextDate().getUTCDate()).toBe(2)
      })
    })
```

File: frontend/src/utils/TimeWithOffset.js (hunk 3 above)
Comment (holgerkoser) [reply frontend/src/utils/TimeWithOffset.js:50]: The method returns the next date in the future with the same time as given by the object.

Comment (holgerkoser) [reply frontend/src/utils/TimeWithOffset.js:50]: It is not sufficient to add 1 day only once!

File: frontend/tests/unit/utils.TimeWithOffset.spec.js (hunk 6 above)
Comment (holgerkoser) [reply frontend/tests/unit/utils.TimeWithOffset.spec.js:43]: Validating only the length of the returned string does not make much sense.

File: frontend/src/components/ShootMessages/ShootMessages.vue [hunk 7]
```diff
@@ -248,34 +259,71 @@ export default {
...
-      if (this.overallStatus === 'info') {
+      if (this.overallSeverity === 'info' || this.overallSeverity === 'verbose') {
         return 'mdi-information-outline'
       }
       return 'mdi-alert-circle-outline'
     },
-    overallStatus () {
-      if (some(this.shootMessages, { color: 'error' })) {
+    overallSeverity () {
+      if (some(this.shootMessages, { severity: 'error' })) {
```
Comment (holgerkoser) [inline frontend/src/components/ShootMessages/ShootMessages.vue:299]:
```suggestion
      if (some(this.shootMessages, ['severity', 'error'])) {
```

File: frontend/src/components/ShootMessages/ShootMessages.vue [hunk 8]
```diff
@@ -248,34 +259,71 @@ export default {
...
         return 'error'
       }
-      if (some(this.shootMessages, { color: 'warning' })) {
+      if (some(this.shootMessages, { severity: 'warning' })) {
```
Comment (holgerkoser) [inline frontend/src/components/ShootMessages/ShootMessages.vue:302]:
```suggestion
      if (some(this.shootMessages, ['severity', 'warning'])) {
```

File: frontend/src/components/ShootMessages/ShootMessages.vue [hunk 9]
```diff
@@ -248,34 +259,71 @@ export default {
...
         return 'warning'
       }
+      if (some(this.shootMessages, { severity: 'verbose' })) {
```
Comment (holgerkoser) [inline frontend/src/components/ShootMessages/ShootMessages.vue:305]:
```suggestion
      if (some(this.shootMessages, ['severity', 'verbose'])) {
```

File: frontend/src/components/ShootMessages/ShootMessages.vue (hunk 7 above)
Comment (holgerkoser) [inline frontend/src/components/ShootMessages/ShootMessages.vue:299]: or maybe even better
```js
  for (const { severity } of this.shootMessages) {
    if (['error', 'warning', 'verbose'].includes(severity)) {
      return severity
    }
  }
  return 'info'
```

**Review by petersutter — APPROVED** [review]
/lgtm

**Review by holgerkoser — APPROVED** [review]
/lgtm

**Pull request merged**